RecoTracker/FinalTrackSelectors/plugins/TrackLwtnnClassifier.cc

```diff
@@ -0,0 +1,124 @@
+{7,0.0},			//detachedTripletStep
+{8,-0.8},			//mixedTripletStep
+{9,-0.6},			//pixelLessStep
+{10,-0.4},			//tobTecStep
+{11,0.6},			//jetCoreRegionalStep
+{22,0.8},			//highPtTripletStep
+{23,0.2},			//lowPtQuadStep
+{24,-0.6},			//detachedQuadStep
+};
+
+namespace {
+  struct lwtnn {
+    lwtnn(const edm::ParameterSet& cfg):
+      lwtnnLabel_(cfg.getParameter<std::string>("lwtnnLabel"))
+    {}
+
+    static const char *name() { return "TrackLwtnnClassifier"; }
+
+    static void fillDescriptions(edm::ParameterSetDescription& desc) {
+      desc.add<std::string>("lwtnnLabel", "trackSelectionLwtnn");
+    }
+
+    void beginStream() {}
+    void initEvent(const edm::EventSetup& es) {
+      edm::ESHandle<lwt::LightweightNeuralNetwork> lwtnnHandle;
+      es.get<TrackingComponentsRecord>().get(lwtnnLabel_, lwtnnHandle);
+      neuralNetwork_ = lwtnnHandle.product();
+    }
+
+    float operator()(reco::Track const & trk,
+                     reco::BeamSpot const & beamSpot,
+                     reco::VertexCollection const & vertices,
+                     lwt::ValueMap & inputs) const {
+      // lwt::ValueMap is typedef for std::map<std::string, double>
+      //
+      // It is cached per event to avoid constructing the map for each
+      // track while keeping the operator() interface const.
+
+      Point bestVertex = getBestVertex(trk,vertices);
+
+      inputs["trk_pt"] = trk.pt();
+      inputs["trk_eta"] = trk.eta();
+      inputs["trk_lambda"] = trk.lambda();
+      inputs["trk_dxy"] = trk.dxy(beamSpot.position()); // Training done without taking absolute value
+      inputs["trk_dz"] = trk.dz(beamSpot.position()); // Training done without taking absolute value
+      inputs["trk_dxyClosestPV"] = trk.dxy(bestVertex); // Training done without taking absolute value
+      inputs["trk_dzClosestPVNorm"] = std::max(-0.2, std::min(trk.dz(bestVertex), 0.2)); // Training done without taking absolute value
+      inputs["trk_ptErr"] = trk.ptError();
+      inputs["trk_etaErr"] = trk.etaError();
+      inputs["trk_lambdaErr"] = trk.lambdaError();
+      inputs["trk_dxyErr"] = trk.dxyError();
+      inputs["trk_dzErr"] = trk.dzError();
+      inputs["trk_nChi2"] = trk.normalizedChi2();
+      inputs["trk_ndof"] = trk.ndof();
+      inputs["trk_nInvalid"] = trk.hitPattern().numberOfLostHits(reco::HitPattern::TRACK_HITS);
+      inputs["trk_nPixel"] = trk.hitPattern().numberOfValidPixelHits();
+      inputs["trk_nStrip"] = trk.hitPattern().numberOfValidStripHits();
+      inputs["trk_nPixelLay"] = trk.hitPattern().pixelLayersWithMeasurement();
+      inputs["trk_nStripLay"] = trk.hitPattern().stripLayersWithMeasurement();
+      inputs["trk_n3DLay"] = (trk.hitPattern().numberOfValidStripLayersWithMonoAndStereo()+trk.hitPattern().pixelLayersWithMeasurement());
+      inputs["trk_nLostLay"] = trk.hitPattern().trackerLayersWithoutMeasurement(reco::HitPattern::TRACK_HITS);
+      inputs["trk_algo"] = trk.algo(); // eventually move to originalAlgo
```

Comment:
what is the implication of the comment ```eventually move to originalAlgo```?
Do the tracks at this point have ```algo != originalAlgo ``` ?

Reply:
This comment has been there for a while, and to be honest I can't recall exactly why. I think it should be possible to have the case of algo != originalAlgo for tracks that have been reconstructed but rejected by an earlier iteration. Probably the comment was there that this is something to be tested at some point, to see what the effect of using originalAlgo instead would have.

However since in the training for this version of the classifier trk_algo was used, I removed the comment as irrelevant.

Reply:
At the stage where the track classification is usually run within an iteration the `algo` is indeed equivalent to `originalAlgo`. But I think it would be a tiny bit more general, less confusing, and less prone to errors in the training to use `originalAlgo` in all stages of the training+inference.